GLHF.chat, and Hyperbolic API's, as well as Gemini-2.0-thinking support, and Deepseek-R1 Ollama support

Sweaterdog opened the pull request:
Adds 2 new API's, a fix for googles latest models, and a fix for Deepseek-R1 support.

GLHF.chat was free, and can host models from huggingface, Hyperbolic Is free and hosts hand-picked models, like Deepseek-R1

Google released Gemini-2.0-flash-thinking, which wasn't supported by the old gemini.js file.

Deepseek-R1 uses specific reasoning tags (<think> and </think>) Which needs a specific method so the commands the model is reasoning with does not become part of the response.

Sweaterdog commented:
Changes made:
Allowed Deepseek-r1 support through the following providers: Huggingface, Hyperbolic, Groqcloud, and Ollama
Made changes to Gemini.js to allow the new thinking model (The endpoint in node_modules/@google/generative-ai/dist/index.js still needs to be changed from "v1" to "v1beta")

uukelele-scratch commented:
I'm getting "SyntaxError: Unexpected token 'else'" (line 60 of prompter.js).

Here are lines 50-81, src/agent/prompter.js.
```
if (typeof chat === 'string' || chat instanceof String) {
            chat = {model: chat};
            if (chat.model.includes('gemini'))
                chat.api = 'google';
            else if (chat.model.includes('gpt') || chat.model.includes('o1'))
                chat.api = 'openai';
            else if (chat.model.includes('claude'))
                chat.api = 'anthropic';
            else if (chat.model.includes('huggingface/'))
                chat.api = "huggingface";
            else if (chat.model.includes('hf:'))
                chat.api = "glhf";
            else if (chat.model.includes('hyperbolic:')|| chat.model.includes('hb:'))
                chat.api = "hyperbolic";
            else if (chat.model.includes('meta/') || chat.model.includes('mistralai/') || chat.model.includes('replicate/'))
            else if (chat.model.includes('meta/') || chat.model.includes('replicate/')) // why's this here?
                chat.api = 'replicate';
            else if (chat.model.includes('mistralai/') || chat.model.includes("mistral/"))
                chat.api = 'mistral';
            else if (chat.model.includes("groq/") || chat.model.includes("groqcloud/"))
                chat.api = 'groq';
            else if (chat.model.includes('novita/'))
                chat.api = 'novita';
             else if (chat.model.includes('qwen'))
                 chat.api = 'qwen';
            else if (chat.model.includes('grok'))
                chat.api = 'xai';
            else if (chat.model.includes('deepseek'))
                chat.api = 'deepseek';
            else
                chat.api = 'ollama';
        }
```

The line
`else if (chat.model.includes('meta/') || chat.model.includes('replicate/'))`
should be removed.

MaxRobinsonTheGreat commented:
sorry, I just did a refactor to models/prompter and gemini specifically. if you merge with main I will take a look

Sweaterdog commented:
Boom, I fixed all of the merge issues and this should now be up to date!

uukelele-scratch commented:
> (The endpoint in node_modules/@google/generative-ai/dist/index.js still needs to be changed from "v1" to "v1beta")

In `package.json`, you can see:
```
...
"scripts": {
        "postinstall": "patch-package",
        "start": "node main.js"
    },
...
```

Perhaps you could add a patch that changed v1 to v1beta in `node_modules/@google/generative-ai/dist/index.mjs`

Sweaterdog commented:
Thanks for reminding me! I will fix that ASAP

Sweaterdog commented:
Okay, I added the patch file to the code repository

Sweaterdog commented:
Fixed the reasoning handling for some models, and fixed the google patch

Sweaterdog commented:
This Pull Request is fully ready to be merged to the latest version.

reformstudios commented on src/models/prompter.js:
```diff
@@ -129,6 +131,10 @@ export class Prompter {
                 model_profile.api = 'mistral';
             else if (profile.model.includes("groq/") || profile.model.includes("groqcloud/"))
                 profile.api = 'groq';
+            else if (chat.model.includes('hf:'))
```
I'm getting an error here. Does chat need to be "profile" here? 

MindServer running on port 8080
Client connected
Registering agents: andy
Agents registered
Starting agent with profile: ./andy.json
Starting agent initialization with profile: ./andy.json
Initializing action manager...
Initializing prompter...
Agent start failed with error
[ReferenceError: chat is not defined
  at Prompter._selectAPI (file:///D:/Games/Minecraft/ai/src/mindcraft/src/models/prompter.js:134:18)
  at new Prompter (file:///D:/Games/Minecraft/ai/src/mindcraft/src/models/prompter.js:59:39)
  at Agent.start (file:///D:/Games/Minecraft/ai/src/mindcraft/src/agent/agent.js:33:29)
  at file:///D:/Games/Minecraft/ai/src/mindcraft/src/process/init_agent.js:58:21
  at file:///D:/Games/Minecraft/ai/src/mindcraft/src/process/init_agent.js:64:3
  at ModuleJob.run (node:internal/modules/esm/module_job:271:25)
  at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:547:26)
  at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:98:5)]
Failed to start agent process:
[ReferenceError: chat is not defined
  at Prompter._selectAPI (file:///D:/Games/Minecraft/ai/src/mindcraft/src/models/prompter.js:134:18)
  at new Prompter (file:///D:/Games/Minecraft/ai/src/mindcraft/src/models/prompter.js:59:39)
  at Agent.start (file:///D:/Games/Minecraft/ai/src/mindcraft/src/agent/agent.js:33:29)
  at file:///D:/Games/Minecraft/ai/src/mindcraft/src/process/init_agent.js:58:21
  at file:///D:/Games/Minecraft/ai/src/mindcraft/src/process/init_agent.js:64:3
  at ModuleJob.run (node:internal/modules/esm/module_job:271:25)
  at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:547:26)
  at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:98:5)]
Agent process exited with code 1 and signal null
Agent process ./andy.json exited too quickly and will not be restarted.

Sweaterdog commented:
Fixed the typos in prompter.js, should be working now

pmacdonald-falcons commented on src/models/local.js:
```diff
@@ -12,43 +12,84 @@ export class Local {
     async embed(text) {
         let model = this.model_name || 'nomic-embed-text';
-        let body = {model: model, prompt: text};
+        let body = { model: model, prompt: text };
```
Prompt should be 'input' in the latest ollama API. https://ollama.com/blog/embedding-models

Sweaterdog commented:
Just reviewed the current Pull Request, and everything works, it is now, guaranteed to be ready to be merged

Sweaterdog commented:
I just merged with the main branch, and now the pull request is ready to be merged

Sweaterdog commented:
Reasons why this should be merged next:

Adds more APIs, so more people can use it

Enables Gemini 2.0 models, which are smarter than the 1.5 models, faster, and work better than gpt-4o

Enables ALL local models to be used, not just llama3

Enables access for reasoning models, such as DeepSeek-R1 (Via any API besides DeepSeek, including ollama) as well as Gemini-2.0-flash-thinking

Fixed the README to tell the user what API keys go where for what API service (currently broken)

This PR has been up for a long time (actually, it's been 2 pull requests)

uukelele-scratch commented:
@Sweaterdog 

Suggestion: In `src/models/prompter.js`, replace the _selectAPI function with this:

```javascript
_selectAPI(profile) {
        if (typeof profile === 'string' || profile instanceof String) {
            profile = {model: profile};
        }
        if (!profile.api) {
            if (profile.model.includes('openrouter/'))
                profile.api = 'openrouter'; // must do first because shares names with other models
            else if (profile.model.includes('together/'))
                profile.api = 'together'; // NOTE: you can remove this because it's only part of my PR that adds together.ai support
            else if (profile.model.includes('ollama/'))
                profile.api = 'ollama'; // also must do early because shares names with other models
            else if (profile.model.includes('gemini'))
                profile.api = 'google';
            else if (profile.model.includes('gpt') || profile.model.includes('o1')|| profile.model.includes('o3'))
                profile.api = 'openai';
            else if (profile.model.includes('claude'))
                profile.api = 'anthropic';
            else if (profile.model.includes('huggingface/'))
                profile.api = "huggingface";
            else if (profile.model.includes('replicate/'))
                profile.api = 'replicate';
            else if (profile.model.includes('mistralai/') || profile.model.includes("mistral/"))
                model_profile.api = 'mistral';
            else if (profile.model.includes("groq/") || profile.model.includes("groqcloud/"))
                profile.api = 'groq';
            else if (profile.model.includes('novita/'))
                profile.api = 'novita';
            else if (profile.model.includes('qwen'))
                profile.api = 'qwen';
            else if (profile.model.includes('grok'))
                profile.api = 'xai';
            else if (profile.model.includes('deepseek'))
                profile.api = 'deepseek';
            else 
                profile.api = 'ollama';
        }
        return profile;
    }
```

And replace _createModel with this:

```javascript
    _createModel(profile) {
        let model = null;
        if (profile.api === 'google')
            model = new Gemini(profile.model, profile.url, profile.params);
        else if (profile.api === 'openai')
            model = new GPT(profile.model, profile.url, profile.params);
        else if (profile.api === 'anthropic')
            model = new Claude(profile.model, profile.url, profile.params);
        else if (profile.api === 'replicate')
            model = new ReplicateAPI(profile.model.replace('replicate/', ''), profile.url, profile.params);
        else if (profile.api === 'ollama')
            model = new Local(profile.model.replace('ollama/', ''), profile.url, profile.params);
        else if (profile.api === 'mistral')
            model = new Mistral(profile.model, profile.url, profile.params);
        else if (profile.api === 'groq')
            model = new GroqCloudAPI(profile.model.replace('groq/', '').replace('groqcloud/', ''), profile.url, profile.params);
        else if (profile.api === 'huggingface')
            model = new HuggingFace(profile.model, profile.url, profile.params);
        else if (profile.api === 'novita')
            model = new Novita(profile.model.replace('novita/', ''), profile.url, profile.params);
        else if (profile.api === 'qwen')
            model = new Qwen(profile.model, profile.url, profile.params);
        else if (profile.api === 'xai')
            model = new Grok(profile.model, profile.url, profile.params);
        else if (profile.api === 'deepseek')
            model = new DeepSeek(profile.model, profile.url, profile.params);
        else if (profile.api === 'openrouter')
            model = new OpenRouter(profile.model.replace('openrouter/', ''), profile.url, profile.params);
        else if (profile.api === 'together')
            model = new Together(profile.model.replace('together/', ''), profile.url, profile.params); // can be removed also
        else
            throw new Error('Unknown API:', profile.api);
        return model;
    }
```

What this does is allow you to specify, e.g.
```ollama/deepseek-r1```
and it knows automatically it's an ollama model, so it goes to ollama, and it only feeds
```deepseek-r1``` as the model to ollama.

MaxRobinsonTheGreat reviewed (changes requested):
I see many issues. Also, there are lots of random formatting changes (adding/removing newlines). Please look at the diffs and remove unnecessary ones, they bloat the PR.

uukelele-scratch commented on src/models/prompter.js:
```diff
@@ -40,7 +42,7 @@ export class Prompter {
         }
         // base overrides default, individual overrides base
-
+        // Removed a bit of space that was right here by adding a comment instead of deleting it because I am making a pull request to this code and I can do whatever I want because I decided to add 2 new API services to Mindcraft now look at me go! Woohoo! I am flying off the edge of the screen oh no!
```
why is this so unnecessarily funny

uukelele-scratch commented on profiles/glhf.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "name": "Good_luck_have_fun",
```
oh better change my together.ai pull request then

MaxRobinsonTheGreat commented on src/models/prompter.js:
```diff
@@ -136,6 +140,10 @@ export class Prompter {
                 profile.api = 'groq';
+            else if (profile.model.includes("GLHF/")
+                profile.api = 'glhf';
+            else if (profile.model.includes("Hyperbolic/")
```
prefixes need to be lower case

MaxRobinsonTheGreat commented on src/models/hyperbolic.js:
```diff
@@ -0,0 +1,114 @@
+// hyperbolic-no-logger.js
+import { getKey } from '../utils/keys.js';
+
+export class hyperbolic {
```
still not changed.

MaxRobinsonTheGreat commented on src/models/prompter.js:
```diff
@@ -39,7 +41,7 @@ export class Prompter {
         }
         // base overrides default, individual overrides base
-
+        // Removed this comment
```
-_-

MaxRobinsonTheGreat commented on settings.js:
```diff
@@ -23,6 +23,8 @@ export default
         // "./profiles/mistral.json",
         // "./profiles/deepseek.json",
+        // "./profiles/glhf.json",
+        // "./profiles/hyperbolic.json",
```
remove

MaxRobinsonTheGreat commented on patches/@google+generative-ai+0.2.1.patch:
```diff
@@ -0,0 +1,12 @@
++++ b/node_modules/@google/generative-ai/dist/index.js
+@@ -156,1 +156,1 @@
+-const API_VERSION = "v1";
++const API_VERSION = "v1beta";
```
I don't want the google patch, why did you resolve this? will it not work with the patch?

uukelele-scratch replied:
> I don't want the google patch, why did you resolve this? will it not work without the patch? 

You need it for gemini-2.0-flash-thinking-exp and other experimental models.

MaxRobinsonTheGreat replied:
this is just bound to be outdated in a month

Sweaterdog replied:
I will remove the google patch, and keep the thinking detection in for now and if google changes it, I can fix it.

MaxRobinsonTheGreat commented on README.md:
```diff
@@ -38,22 +38,26 @@ You can configure the agent's name, model, and prompts in their profile like `an
+| Deepseek | `DEEPSEEK_API_KEY` | `deepseek-reasoner` | [docs](https://api-docs.deepseek.com/) |
+| Openrouter | `OPENROUTER_API_KEY` | `openrouter/anthropic/claude-3.5-sonnet` | [docs](https://openrouter.ai/models) |
+| glhf.chat | `GHLF_API_KEY` | `glhf/hf:meta-llama/Llama-3.1-405B-Instruct` | [docs](https://glhf.chat/user-settings/api) |
+| Hyperbolic | `HYPERBOLIC_API_KEY` | `hyperbolic/deepseek-ai/DeepSeek-V3` | [docs](https://docs.hyperbolic.xyz/docs/getting-started) |
```
you've overwritten the entire table, API names are now wrong. Keep the table as it is, just add the two new APIs.
also, is `glhf/hf:meta-llama/Llama-3.1-405B-Instruct` supposed to have the `hf:` still?

MaxRobinsonTheGreat commented on README.md:
```diff
@@ -38,22 +38,26 @@ You can configure the agent's name, model, and prompts in their profile like `an
 If you use Ollama, to install the models used by default (generation and embedding), execute the following terminal command:
-`ollama pull llama3 && ollama pull nomic-embed-text`
+`ollama pull llama3.1 && ollama pull nomic-embed-text`
+Note that you can you any local model, such as `qwen2.5` `llama3.1` or fine tuned models like `sweaterdog/andy-3.6` But not matter what model, you must put `ollama/` before the model so Mindcraft can handle it properly.
```
this is already clear in the example model name above, lots of APIs require the prefix. remove this comment

MaxRobinsonTheGreat commented on evaluation_script.py:
```diff
@@ -347,4 +347,4 @@ def main():
 if __name__ == "__main__":
-    main()
\ No newline at end of file
+    main()
```
I still see like 5 random newline addition/removal. again, please clean it up. look at the diff.

MaxRobinsonTheGreat reviewed (changes requested):
mo problems. I don't think this will work as is, the class names and prefixes are inconsistent. 

Please test and check the diff before requesting review.

Sweaterdog commented:
I reviewed all of the files, and made the requested changes. I also removed unnecessary comments and extra spaces on files that I made, I haven't touched any other files.

MaxRobinsonTheGreat commented:
@Sweaterdog bro, look at the diff here on github. you are resolving conversations on unchanged files. very frustrating.

Sweaterdog commented:
> @Sweaterdog bro, look at the diff here on github. you are resolving conversations on newlines that you have in fact changed. very frustrating.

I'm sorry I have been a problem, however on my end when looking at the diffs it says everything you mentioned has been fixed, hence why I resolved it. Just to cover what I did:

Fixed the export class for Hyperbolic, hyperbolic-> Hyperbolic, glhf -> GLHF

Removed useless comments and spaces from files I have made/ edited
Prompter.js long comment -> Gone
GLHF.js -> Kept comments and removed the unnecessary clarification of what file at the top
Hyperbolic.js -> removed the extra white space at the bottom of the file 
Google patch -> I did not resolve, I can remove it if absolutely necessary, but most of this PR won't work without that patch
Extra spaces -> I've gone through all of the files and removed unnecessary spaces

I will merge with the main whenever I am able to, for now, I will convert this PR to a draft in case anything else needs fixed.

Again, apologies for being a nuisance. This is my first PR, ever, so again I am sorry if anything I am doing is considered "illegal" in the coding space.

MaxRobinsonTheGreat commented:
apologies for being rude. You won't see these differences locally, you can see the diff [here](https://github.com/kolbytn/mindcraft/pull/423/files). You need to go through and manually add/remove the spaces and newlines. Looks like [he fixed it](https://github.com/kolbytn/mindcraft/pull/423/files) for you if you merge this in, but there are conflicts now

This is what I'm talking about:

<img width="621" alt="Capture" src="https://github.com/user-attachments/assets/575f0706-f3ca-4cfe-b999-0ac0b46672e8" />

Sweaterdog commented:
> apologies for being rude. You won't see these differences locally, you can see the diff [here](https://github.com/kolbytn/mindcraft/pull/423/files). You need to go through and manually add/remove the spaces and newlines. Looks like [he fixed it](https://github.com/kolbytn/mindcraft/pull/423/files) for you if you merge this in, but there are conflicts now
> 
> This is what I'm talking about:
> 
> <img alt="Capture" width="621" src="https://private-user-images.githubusercontent.com/9001247/420995041-575f0706-f3ca-4cfe-b999-0ac0b46672e8.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE2MzEwNjUsIm5iZiI6MTc0MTYzMDc2NSwicGF0aCI6Ii85MDAxMjQ3LzQyMDk5NTA0MS01NzVmMDcwNi1mM2NhLTRjZmUtYjk5OS0wYWMwYjQ2NjcyZTguUE5HP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMTBUMTgxOTI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTllOTIzMmEzZjk5N2Y1MjgzMTRhNjIzMzM3MDY5OTVlNmIxYzFlN2NiOTE4NmQ0YzFjZjQwMGJmNjIwYTcyYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.JGkcTKBI8vzYECLJycYsJ5IIcsS3BOJjIFbG6NSOu9g">

You are fine, It seems when I try to fix the extra space issue on my phone, github automatically adds an extra line, I will try to fix it when I am able to use my computer again.

MaxRobinsonTheGreat commented:
aight im just gonna merge into develop

MaxRobinsonTheGreat commented:
in the future be more mindful of formatting changes. it generates more work for me.

Sweaterdog commented:
> in the future be more mindful of formatting changes. it generates more work for me.

I'm sorry I didn't fix the formatting stuff. I've been away from home and I was unable to fix the issues on my phone, as Github would add an extra line every time I pushed a commit, which was frustrating for me since I fixed it multiple times, but not really changes have been made. I promise I was not trying to make things hard on you.

Munzer1174 commented:
Expert, I've locally deployed the DeepSeek-R1 model. Could you provide specific guidance on how to implement it in-game?Thank you very much.（My English isn't very good, so I apologize in advance for any mistakes.)    :）

Sweaterdog commented:
In order to use the local DeepSeek-R1 model, you need to specify the exact name of the model, with ollama/ before it. For the default 7b model (which is a math reasoning model) you need to set `ollama/deepseek-r1` as the model.

Munzer1174 commented:
I set the model in andy.json to ollama/deepseek-r1 (I have the 7B version deployed locally), but I still encountered the following error when running the command line.【I have run the ollama3.1 model locally and redeployed both the mindcraft and deepseek-r1 7B versions following the README instructions. Could you help me figure out what's going on? I really appreciate you taking the time to respond.】

D:\Games\Minecraft\mindcraft main ai\mindcraft\mindcraft-main updated>node main.js
[ './andy.json' ]
MindServer running on port 8080
Client connected
Registering agents: ColdWall
Agents registered
Starting agent with profile: ./andy.json
Starting agent initialization with profile: ./andy.json
Initializing action manager...
Initializing prompter...
Failed to start agent process:
API key "DEEPSEEK_API_KEY" not found in keys.json or environment variables!
Error: API key "DEEPSEEK_API_KEY" not found in keys.json or environment variables!
  at getKey (file:///D:/Games/Minecraft/mindcraft%20main%20ai/mindcraft/mindcraft-main%20updated/src/utils/keys.js:17:15)
  at new DeepSeek (file:///D:/Games/Minecraft/mindcraft%20main%20ai/mindcraft/mindcraft-main%20updated/src/models/deepseek.js:13:25)
  at Prompter._createModel (file:///D:/Games/Minecraft/mindcraft%20main%20ai/mindcraft/mindcraft-main%20updated/src/models/prompter.js:182:21)
  at new Prompter (file:///D:/Games/Minecraft/mindcraft%20main%20ai/mindcraft/mindcraft-main%20updated/src/models/prompter.js:58:32)
  at Agent.start (file:///D:/Games/Minecraft/mindcraft%20main%20ai/mindcraft/mindcraft-main%20updated/src/agent/agent.js:32:25)
  at file:///D:/Games/Minecraft/mindcraft%20main%20ai/mindcraft/mindcraft-main%20updated/src/process/init_agent.js:58:21
  at file:///D:/Games/Minecraft/mindcraft%20main%20ai/mindcraft/mindcraft-main%20updated/src/process/init_agent.js:65:3
  at ModuleJob.run (node:internal/modules/esm/module_job:268:25)
  at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:543:26)
  at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:116:5)
Agent process exited with code 1 and signal null
Agent process ./andy.json exited too quickly and will not be restarted.
 


[andy.json](https://github.com/user-attachments/files/19486835/andy.json)

Sweaterdog commented:
I would highly suggest going to the discord server. I speak English and some Spanish, however others are fluent in other languages and can help you.

Munzer1174 commented:
ok,i will try it,thank you for your respond.
